pkg/auth/oauthserver/metadata.go

```diff
@@ -0,0 +1,89 @@
+package oauthserver
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/lestrrat-go/jwx/jwk"
+
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/flyteorg/flyteadmin/pkg/auth/interfaces"
+)
+
+var (
+	tokenRelativeUrl             = mustParseURL("/oauth2/token")
+	authorizeRelativeUrl         = mustParseURL("/oauth2/authorize")
+	authorizeCallbackRelativeUrl = mustParseURL("/oauth2/authorize_callback")
+	jsonWebKeysUrl               = mustParseURL("/oauth2/jwks")
+)
+
+func mustParseURL(rawURL string) *url.URL {
+	res, err := url.Parse(rawURL)
+	if err != nil {
+		panic(err)
+	}
+
+	return res
+}
+
+type DiscoveryDocument struct {
+	Issuer                            string   `json:"issuer"`
+	AuthorizationEndpoint             string   `json:"authorization_endpoint"`
+	TokenEndpoint                     string   `json:"token_endpoint"`
+	ResponseTypesSupported            []string `json:"response_types_supported"`
+	ScopesSupported                   []string `json:"scopes_supported"`
+	TokenEndpointAuthMethodsSupported []string `json:"token_endpoint_auth_methods_supported"`
+	JSONWebKeysUri                    string   `json:"jwks_uri"`
+	CodeChallengeMethodsSupported     []string `json:"code_challenge_methods_supported"`
+
+	// optional
+	GrantTypesSupported []string `json:"grant_types_supported"`
+}
+
+// GetJSONWebKeysEndpoint serves requests to the jwks endpoint.
+// ref: https://tools.ietf.org/html/rfc7517
+func GetJSONWebKeysEndpoint(authCtx interfaces.AuthenticationContext) http.HandlerFunc {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		s := authCtx.OAuth2Provider().KeySet()
+		raw, err := json.Marshal(s)
+		if err != nil {
+			http.Error(writer, fmt.Errorf("failed to write public key. Error: %w", err).Error(), http.StatusInternalServerError)
+			return
+		}
+
+		writer.Header().Set("Content-Type", "application/json")
+		size, err := writer.Write(raw)
+		if err != nil {
+			logger.Errorf(context.Background(), "Wrote JSONWebKeys response size %d, err %s", size, err)
+		}
+	}
+}
+
+func getJSONWebKeys(publicKeys []rsa.PublicKey) (jwk.Set, error) {
+	s := jwk.NewSet()
+	for _, publicKey := range publicKeys {
+		key, err := jwk.New(publicKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write public key. Error: %w", err)
+		}
+
+		err = jwk.AssignKeyID(key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write public key. Error: %w", err)
+		}
+
+		err = key.Set(KeyMetadataPublicCert, &publicKey)
```

Comment:
Passing this as a pointer is causing overwrite of the KeyMetadataPublicCert of the first key during the next iteration .

Reply:
this can happen if publicKey is an interface type?
Else that makes no sense.
But in anycase if you want to take a reference of a loop var, it is a good practice to assign it to a loop local variable.

Reply:
The set function is indeed accepting an interface
`Set(string, interface{}) error`

I think that solved the issue i was facing with authentication with rotated keys by doing the suggested.
```
var localPublicKey rsa.PublicKey
		localPublicKey = publicKey
		err = key.Set(KeyMetadataPublicCert, &localPublicKey)
```

There are two publickeys this loop iterates on , 0th being the new one and 1st being the old one.
On the first iteration
key0 -> &publickey0
On the second iteration as the loop variable publickey is initialised with publickey1
The pointer for publicKey0 is pointing to publickey1's data

And in the end 
key0 -> pointing to publicKey1data
key1 -> pointing to publicKey1data

The new token seems to be signed by publickey0data .(Since we want to issue token with new key only)
And then verifiy using the publicKey1data which is old key and was causing an issue on my end where  the auth flow would complete and succeed but when calling api with the issued token would fail in key verification stage.

Reply:
Created this MR Created this MR https://github.com/flyteorg/flyteadmin/pull/183/